src/manager/manager.go

```diff
@@ -0,0 +1,230 @@
+                }
+        }
+
+        // Recursively search each directory (DFS)
+        // After file processing to improve spatial locatlity for files
+        for _, subDir := range dirs {
+                z.WalkDir(path.Join(dir, subDir), subDir, false)
+        }
+
+        // root dir not marked as directory
+        if !root {
+                z.IncludeFolderEnd()
+        }
+}
+
+// TODO: Change to interface for Metadata to have diff types of Metadata
+// IncludeFolderBegin implements Manager.IncludeFolderBegin
+func (z *ZarManager) IncludeFolderBegin(name string) {
+        h := &FileMetadata{
+                        Begin   : -1,
+                        End     : -1,
+                        Name    : name,
+                        Type    : Directory,
+        }
+
+        // Add to the image's Metadata at end
+        z.Metadata = append(z.Metadata, *h)
+}
+
+// IncludeFolderEnd implements IncludeFolderEnd
+func (z *ZarManager) IncludeFolderEnd() {
+        h := &FileMetadata{
+                        Begin   : -1,
+                        End     : -1,
+                        Name    : "..",
+                        Type    : Directory,
+        }
+
+        // Add to the image's Metadata at end
+        z.Metadata = append(z.Metadata, *h)
+}
+
+// IncludeSymlink adds Metadata to the image file for a symbolic link. This
+// allows for paths to be indirections. Not included in interface because
+// not necessarily fundamental for correctness.
+//
+// parameter (name)     : name of file
+// parameter (link)     : the actual path to the desired file
+func (z *ZarManager) IncludeSymlink(name string, link string) {
+        h := &FileMetadata{
+                        Begin   : -1,
+                        End     : -1,
+                        Name    : name,
+                        Link    : link,
+                        Type    : Symlink,
+        }
+        z.Metadata = append(z.Metadata, *h)
+}
+
+// IncludeFile implements Manager.IncludeFile
+func (z *ZarManager) IncludeFile(fn string, basedir string) (int64, error) {
+        content, err := ioutil.ReadFile(path.Join(basedir, fn))
+        if err != nil {
+                log.Fatalf("can't include file %v, err: %v", fn, err)
+                return 0, nil
+        }
+
+        // Retrieve the current offset into the file and write the file contents
+        oldCounter := z.Writer.Count
+        real_end, err := z.Writer.Write(content, z.PageAlign)
+        if err != nil {
+                        log.Fatalf("can't write to file")
+                        return 0, err
+        }
+
+        // Create the file Metadata
+        h := &FileMetadata{
+                        Begin   : oldCounter,
+                        End     : real_end,
+                        Name    : fn,
+                        Type    : RegularFile,
+        }
+        z.Metadata = append(z.Metadata, *h)
+
+        return real_end, err
+}
+
+// TODO: Is gob the best choice here?
```

Comment:
I also feel like gob is not the best choice here but I didn't find a better replacement. Please feel free to suggest any better options.

Reply:
I will look into it and let you know. I am looking for something that supports direct conversion to inodes within gvisor. I will let you know if I find anything, otherwise this is a good option.

Reply:
With gob it is definitely a good idea.
Some information (sorry I didn't write comments in imgfs side)

inode can be:
1. ramfs directory (there is a internal map in ramfs directory there which can map names to different inodes. That's one way directories work in Sentry.)
2. ramfs symbol link
3. imgfs file

Please try to build the tree in car and see if gob can help you preload pre-constructed tree.

Thanks!